优化 AndroidManifestParser 读取方式

AyakaAgo opened the pull request:
对于 app 自身，直接使用 Context；对于其他已安装的应用，使用 Context.createPackageContext。

保留原有 apkPath 方式。

Commit:
优化 AndroidManifestParser 读取方式

对于 app 自身，直接使用 Context；对于其他已安装的应用，使用 Context.createPackageContext。

保留原有 apkPath 方式。

getActivity commented:
请详细阐述优化带来的好处有哪些？

AyakaAgo commented:
此 pr 的内容是将**已安装的应用**（[自身](https://github.com/AyakaAgo/XXPermissions/blob/e4ed6bb38f03db4962765a27164078efe76c3dee/library/src/main/java/com/hjq/permissions/manifest/AndroidManifestParser.java#L254)和[其他应用](https://github.com/AyakaAgo/XXPermissions/blob/e4ed6bb38f03db4962765a27164078efe76c3dee/library/src/main/java/com/hjq/permissions/manifest/AndroidManifestParser.java#L256)）的 AndroidManifest.xml 读取改为不 `addAssetPath` 的方法

-------------------------------------

`AssetManager.addAssetPath` 会污染 AssetManager 以及相关操作，包括但不限于（按严重程度）：
- `AssetManager.openXxx(  )` 的**同名文件会打开 addAssetPath 包的文件**而不是 app 本身的文件
- `AssetManager.list("")` 目录列表**会混入 addAssetPath 包的文件**

![受影响的 assets](https://github.com/user-attachments/assets/55110500-a346-4c03-b539-4829d3ccb690)

而 [Context.createPackageContext](https://developer.android.com/reference/android/content/Context#createPackageContext(java.lang.String,%20int)) 则不会有此问题，因为 assets 依然隔离。

----------------------------------------------

对于[此库的用例](https://github.com/getActivity/XXPermissions/blob/c64591b27c45a182f3d22f8992ba2858e4a58977/library/src/main/java/com/hjq/permissions/XXPermissions.java#L251)，虽然 [AndroidManifestParser#L81](https://github.com/getActivity/XXPermissions/blob/c64591b27c45a182f3d22f8992ba2858e4a58977/library/src/main/java/com/hjq/permissions/manifest/AndroidManifestParser.java#L81) 将**自身 sourceDir** _再次加入_ assetpath 并不会有上述问题，但对读取自己的 assets 来说，这属于多余的行为（asset path 本身就包含自己的 apk），直接使用 `getAssets().openXmlResourceParser("AndroidManifest.xml")` 即可。

对用外部调用者，由于 [AndroidManifestParser.findApkPathCookie](https://github.com/getActivity/XXPermissions/blob/c64591b27c45a182f3d22f8992ba2858e4a58977/library/src/main/java/com/hjq/permissions/manifest/AndroidManifestParser.java#L107) 是公共方法，`addAssetPath` 会带来上述潜在问题。

---------------------------------

虽然 `addAssetPath` 带来的污染可以“remove”掉，但移除需要更多反射（[AssetManager.getApkAssets()](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/content/res/AssetManager.java?q=symbol%3A%5Cbandroid.content.res.AssetManager.getApkAssets%5Cb%20case%3Ayes)，[ApkAssets.getAssetPath()](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/content/res/ApkAssets.java?q=symbol%3A%5Cbandroid.content.res.ApkAssets.getAssetPath%5Cb%20case%3Ayes)）_（unsupported 仍可访问）_，且不能保证这些 non-sdk 接口在以后的版本仍可访问。

可以新增方法来提供 apk assetpath 移除支持，但由于这并非此库的主要目的，外部调用者不应过于依赖此部分。

getActivity commented:
> AssetManager.addAssetPath 会污染 AssetManager 以及相关操作

* 你怎么知道这样做会污染？虽然反射调用了 addAssetPath，仅仅是为了获得 apk 路径对应的 cookie，但是里面已经有了这个 apk 的路径了，所以一般情况下并不会添加进去，导致污染的根据是什么？

<img width="1489" height="862" alt="image" src="https://github.com/user-attachments/assets/07e1c1d2-c98c-4f70-8775-d61777521441" />

getActivity commented:
> 直接使用 `getAssets().openXmlResourceParser("AndroidManifest.xml")` 即可

* 调用 `openXmlResourceParser` 但是不传 cookie 的方法，这样 cookie 默认会用 0 的值，但是通过 debug 获取到当前应用的 apk 路径对应的 cookie 是 8，这逻辑显然有问题，另外你怎么确保调用 `getAssets().openXmlResourceParser("AndroidManifest.xml")` 不会读错别的 apk 的清单文件？

<img width="1005" height="809" alt="image" src="https://github.com/user-attachments/assets/ab8349f6-05f3-4fbc-a2d3-d11e4269f0e9" />

<img width="1492" height="800" alt="image" src="https://github.com/user-attachments/assets/aa5c8977-30e4-4318-afc8-e44ec0f09fa3" />

Commit:
优化 AndroidManifestParser 读取方式

PR 406 的追加修正

对于 app 自身和其他已安装的应用，都使用 Context.createPackageContext，以获取 raw assets 内容。

保留原有 apkPath 方式。

AyakaAgo commented:
感谢补充

> 你怎么知道这样做会污染？虽然反射调用了 addAssetPath，仅仅是为了获得 apk 路径对应的 cookie，但是里面已经有了这个 apk 的路径了，所以一般情况下并不会添加进去，导致污染的根据是什么？

这**仅是针对于此库的用例**

**对于外部使用者**来说，就如上面所说的 [AndroidManifestParser.findApkPathCookie](https://github.com/getActivity/XXPermissions/blob/c64591b27c45a182f3d22f8992ba2858e4a58977/library/src/main/java/com/hjq/permissions/manifest/AndroidManifestParser.java#L107) 是公共方法，如果有人用**其他 apk 路径**调用了此方法，那就有机会造成这种污染⚠️ _这是作为公共库的考量_

所以说 _可以新增方法来提供 assetpath 移除支持_

> 导致污染的根据是什么?

#102 就是这种污染，因为 [WebView 会往 asset path 加 Chromium 依赖](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/java/android/webkit/WebViewDelegate.java?q=symbol%3A%5Cbandroid.webkit.WebViewDelegate.addWebViewAssetPath%5Cb%20case%3Ayes)，而 Chromium 对这种重名文件冲突也做过[修复](https://issues.chromium.org/issues/357131361)，但这种修复无法涉及 AndroidManifest（已提交 issuetracker）

针对 WebView 的情况<sup>1</sup>，[commit#2][c2] 进行了追加修正，无论是自身还是其他 app **都进行 `Context.createPackageContext(packageName)` 来获取原的 assets**（不包含 WebView 和其他 `addAssetPath`）

相比反射更加“干净”，且耗时也是 ~1 毫秒 - 毫秒级（_various_），不易察觉

> 不传 cookie 的方法，这样 cookie 默认会用 0 的值

根据 [AssetManager 的打开逻辑][am_open] cookie 0 并不是从 index 0 的 `framework-res.apk` 读取，而是[在 paths 中由 index 高到低 _尝试_ 读取](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/libs/androidfw/AssetManager2.cpp;drc=61197364367c9e404c7da6900658f1b16c42d0da;l=607)：
（通常的顺序）
（低 index）系统资源、静态 rro 包（index 0 时跳过）、apk 自身、WebView 资源包（通过 `Context.createPackageContext` 不包含）、\<uses-static-library\><sup>1</sup>、动态 rro 包（index 0 时跳过）、`addAssetPath` 添加的资源包（通过 `Context.createPackageContext` 不包含）（高 index）

即使初始化了 WebView 组件，`Context.createPackageContext` 的 `getAsset` 不会包含比如 WebView 资源包的动态添加的 asset path，虽然 rro 包也包含 manifest，但 [rro 在 index 0 跳过不读取](https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/libs/androidfw/AssetManager2.cpp;drc=61197364367c9e404c7da6900658f1b16c42d0da;l=612)，实际上相当于读取 apk 自身内容，所以在[改为使用 `Context.createPackageContext`](https://github.com/AyakaAgo/XXPermissions/blob/1d78dfc2db820fe764ad530e63c80ccee7a76476/library/src/main/java/com/hjq/permissions/manifest/AndroidManifestParser.java#L265) 后，可以不通过 `addAssetPath`<sup>1</sup>

> 另外你怎么确保调用 getAssets().openXmlResourceParser("AndroidManifest.xml") 不会读错别的 apk 的清单文件？

**基于 [commit#2][c2] 的修改**，`Context.createPackageContext` 的 getAsset 不会包含比如 WebView 资源包的动态添加的 asset path，根据 [AssetManager 的打开逻辑][am_open]读取到的就是自己 apk 下的 assets 和清单文件<sup>1</sup>，而其它应用通过 `Context.createPackageContext` 读到的也是其它 app 自己的清单文件<sup>1</sup>

--------------------------------------

**外部调用者**的步骤：
1. getAssets().openXmlResourceParser("某个文件") -> 自己的 manifest 或其他文件
2. AndroidManifestParser.findApkPathCookie("**其他 apk 路径**")
_或_
`AndroidManifestParser.findApkPathCookie` 或反射 `addAssetPath`("**其他 apk 路径**")`
_或_
初始化 WebView
3. getAssets().openXmlResourceParser("某个文件") -> 非自己的 manifest 或其他重名文件

完成步骤 2 后，此时除非用 `Context.createPackageContext` 或再次自行通过 `addAssetPath`/`findCookieForPath` 查询自身 apk 的 cookie index，外部调用者**读取重名文件时可能会出现问题**⚠️

----------------------------

补充说明

> 优化带来的好处有哪些？

也为了**减少非 SDK 接口反射**，`findCookieForPath` 已经是[屏蔽状态](https://developer.android.com/guide/app-compatibility/restrictions-non-sdk-interfaces?hl=en#list-names)，避免以后 `addAssetPath`（unsupported）在以后的版本亦变为屏蔽状态（标记为 unsupported 的接口：Non-SDK interfaces that are unrestricted and your app can use. Note however, that these interfaces are unsupported and subject to change without notice. **Expect these interfaces to be conditionally blocked in future Android versions in a max-target-x list.**）

需要留意非 SDK 接口限制的更新。

> 外部调用者不应过于依赖此部分

库内部可以继续使用，但：
- 添加文档来提示使用者可能的资源冲突问题
- 添加对 `addAssetPath` 资源的移除方法\
_或_\
停止 public [AndroidManifestParser.findApkPathCookie](https://github.com/getActivity/XXPermissions/blob/c64591b27c45a182f3d22f8992ba2858e4a58977/library/src/main/java/com/hjq/permissions/manifest/AndroidManifestParser.java#L107) 方法

**读取自身的 AndroidManifest [commit#2][c2] 的修改已经满足需求了**<sup>1</sup>。

对于外部调用者、特例 \<uses-static-library\>、避免以后非 SDK 接口无法访问：
- 引入 [AXML 解析器](https://github.com/LibChecker/LibChecker/blob/67390f8470c52a69f24df1f01758bedb0c6ec5aa/app/src/main/kotlin/com/absinthe/libchecker/utils/manifest/HiddenPermissionsReader.java#L27)
- 条件地迁移到使用标准的 PackageInfo/ApplicationInfo

---------------------------------------

1. 特例 \<uses-static-library\>
和使用 WebView 不同的是，\<uses-static-library\> 的 path 会包含于 context.createPackageContext 中，且 \<uses-static-library\> 的 asset index 会比自身 apk 高。\<uses-static-library\> 是没有写入文档的，但声明了也会生效（例：Chrome），如果外部调用者自身有 \<uses-static-library\> 或 findApkPathCookie 了此类 apk，仍需要保留反射路径（见 [commit](https://github.com/AyakaAgo/XXPermissions/blob/199eecc69f7224508cf914df653f30a25dece90b/library/src/main/java/com/hjq/permissions/manifest/AndroidManifestParser.java#L308)）。\
\
但由于 \<uses-static-library\> 并未通过文档公开，假设调用者不使用的情况，可以忽略。但仍保留反射路径。

[c2]: https://github.com/getActivity/XXPermissions/pull/406/commits/1d78dfc2db820fe764ad530e63c80ccee7a76476
[am_open]: https://cs.android.com/android/platform/superproject/main/+/main:frameworks/base/core/jni/android_util_AssetManager.cpp?q=symbol%3A%5Cbandroid%3A%3ANativeOpenXmlAsset%5Cb%20case%3Ayes

getActivity commented:
> 对于外部调用者、特例 <uses-static-library>、避免以后非 SDK 接口无法访问：
>  1. 引入 AXML 解析器
>  2. 条件地迁移到使用标准的 PackageInfo/ApplicationInfo

1. 解析清单文件的功能对于权限框架来讲不是必须的，可有可无，能读就读，读不到就拉倒，况且现在又不是读不了，根本没有必要多引入一个库

2. 你确定能完全替换？下面这些属性，你确定通过系统提供的方法可以获取？

<img width="845" height="743" alt="image" src="https://github.com/user-attachments/assets/d4474031-d946-4cff-a80c-3d70bc8c3177" />

<img width="884" height="627" alt="image" src="https://github.com/user-attachments/assets/a9a65eb0-0e39-480b-84c9-06099261fc63" />

<img width="1040" height="804" alt="image" src="https://github.com/user-attachments/assets/3ec5a3bd-fc50-42a8-bc1e-6512b503c739" />

<img width="678" height="372" alt="image" src="https://github.com/user-attachments/assets/529019db-a0d3-49d7-8804-98ab69a34fd5" />
